docs/FormalVerification.md

```diff
@@ -56,3 +56,266 @@ building formal verification tools based on CIRCT. The
+  %c1_i32 = hw.constant 1 : i32
+  %c2_i32 = hw.constant 2 : i32
+  %c-1_i32 = hw.constant -1 : i32
+  %0 = comb.modu %arg0, %c2_i32 : i32
+  %1 = comb.icmp eq, %0, %c0_i32 : i32
+  verif.assume %1 : i1
+  %2 = comb.modu %rst_val, %c2_i32 : i32
+  %3 = comb.icmp eq, %2, %c0_i32 : i32
+  verif.assume %3 : i1
+  %4 = comb.xor %rst, %true : i1
+  %5 = ltl.delay %4, 2 : i1
+  %6 = ltl.concat %rst, %5 : i1, !ltl.sequence
+  %7 = ltl.clock %6, posedge %clk : !ltl.sequence
+  verif.assume %7 : !ltl.sequence
+  %8 = comb.add %arg0, %state0 : i32
+  %9 = comb.xor %state0, %c-1_i32 : i32
+  %10 = comb.modu %9, %c2_i32 : i32
+  %11 = comb.icmp eq, %10, %c1_i32 : i32
+  verif.assert %11
+  %12 = comb.mux %rst, %rst_val, %8 : i32
+  verif.yield %9, %12 : i32, i32
+}
+```
+
+This CIRCT Core level representation of a BMC task can then be lowered to the
+SMT dialect like the following.
+
+```mlir
+func.func @helper(%cycle: i32, %prev_clk: !smt.bv<1>, %clk: !smt.bv<1>,
+                  %state0: !smt.bv<32>) -> !smt.bv<32> {
+  %zero = smt.bv.constant #smt.bv<0> : !smt.bv<32>
+  %one = smt.bv.constant #smt.bv<1> : !smt.bv<32>
+  %c-1_i32 = smt.bv.constant #smt.bv<-1> : !smt.bv<32>
+  %two = smt.bv.constant #smt.bv<2> : !smt.bv<32>
+  %arg0 = smt.declare_fun "arg0" : !smt.bv<32>
+  %rst = smt.declare_fun "rst" : !smt.bv<1>
+  %rst_val = smt.declare_fun "rst_val" : !smt.bv<32>
+  %smt_cycle = smt.from_concrete %cycle : i32 -> !smt.bv<32>
+  %true = smt.bv.constant #smt.bv<1> : !smt.bv<1>
+  %false = smt.bv.constant #smt.bv<0> : !smt.bv<1>
+  // assumptions for this cycle 
+  %2 = smt.bv.urem %arg0, %two : !smt.bv<32>
+  %3 = smt.eq %2, %zero : !smt.bv<32>
+  %4 = smt.bv.urem %rst_val, %two : !smt.bv<32>
+  %5 = smt.eq %4, %zero : !smt.bv<32>
+  %6 = smt.bv.cmp ult, %smt_cycle, %two : !smt.bv<32>
+  %7 = smt.bv.cmp uge, %smt_cycle, %two : !smt.bv<32>
+  %8 = smt.eq %rst, %true : !smt.bv<1>
+  %9 = smt.and %6, %8
+  %10 = smt.eq %rst, %false : !smt.bv<1>
+  %11 = smt.and %7, %10
+  %12 = smt.or %9, %11
+  %assumption = smt.and %3, %5, %12
+  // circuit lowered to smt
+  %13 = smt.bv.add %arg0, %state0 : !smt.bv<32>
+  %14 = smt.bv.xor %state0, %c-1_i32 : !smt.bv<32>
+  %15 = smt.bv.urem %14, %two : !smt.bv<32>
+  %16 = smt.eq %15, %one : !smt.bv<32>
+  %17 = smt.implies %assumption, %16
+  smt.assert %17
+  %18 = smt.ite %rst, %rst_val, %13 : !smt.bv<32>
+  %19 = smt.bv.not %prev_clk : !smt.bv<1>
+  %posedge = smt.bv.and %clk, %19 : !smt.bv<1>
+  %20 = smt.ite %posedge, %18, %state0 : !smt.bv<32>
+  return %20 : !smt.bv<32>
+}
+
+smt.solver {
+  %state0_init = smt.declare_fun "state0" : !smt.bv<32>
+  scf.for %i = 0 to 10 step 1 iter_args(%state0 = %state0_init) {
+    %clk = smt.bv.constant #smt.bv<0> : !smt.bv<1>
+    %clk_nxt = smt.bv.constant #smt.bv<1> : !smt.bv<1>
+    %new_state0 = func.call @helper(%i, %clk_nxt, %clk, %state0)
+      : (i32, !smt.bv<1>, !smt.bv<1>, !smt.bv<32>) -> !smt.bv<32>
+    smt.check sat {} unknown {} unsat {}
```

Comment:
Are these `smt.check` op regions just left empty because it's curently undecided how to handle assertions being violated? Or is there some way success/failure is signalled that I've missed?